### ISSUE TITLE
Validation retains InfiniteDataLoader workers and prefetched batches between epochs

### ISSUE THREAD

**Issue by danielmorozoff** [ultralytics/ultralytics#25827]
## GPU comparison — 2026-09-06

[Fresh measurements on PR #26014](https://github.com/ultralytics/ultralytics/pull/26014#issuecomment-5555193776) confirm that the current finite/persistent design drains validation queues. The 1920px small-data check retained about 64 MiB less shared memory with identical predictions. Matched 20-epoch training took 4.8% longer on COCO8 and 11.7% longer on COCO128, with identical per-epoch metrics. Recommend no default loader change from these results alone; the original GB10 memory-pressure workload remains unverified.

## Current status — 2026-09-05

- [x] Lower validation prefetch shipped in #26057 (8.4.139): [current `build_dataloader`](https://github.com/ultralytics/ultralytics/blob/5eda95e839a9601fc479f1b5e042d48d7e77fa13/ultralytics/data/build.py) uses **2 for unshuffled loaders**, retaining 4 for shuffled training. The original report below describes 8.4.120 and prefetch 4; its 80-batch/~27-GiB queue estimate is historical (at depth 2, the same raw-image estimate is 40 batches/~13.2 GiB, before overhead).
- [ ] The reported GB10/1920px/batch16 memory-pressure case is **not yet reproduced on current main**. This session's real-image Linux/macOS worker check confirms workers remain alive after a pass and terminate at `close()`, with prefetch 2. It does not reproduce the original 95–100-GB pressure.
- [ ] #26014 remains open. Its current proposal is finite validation iteration with **persistent workers**, not unconditional worker teardown. Validate its current head against the current prefetch-2 baseline on the reported GB10 workload, measuring peak/settled PSS, epoch wall time, and unchanged metrics together.

Preserve the explicit maintainer decision to reuse validation workers during training. Historical matched measurements of the earlier #26014 teardown version showed unconditional teardown/recreation increased wall time ~25.3% and reduced throughput ~20.2%; **recommend no action on that rejected approach**. Later #26014 historical GPU trials were workload-dependent (COCO8 ~5.7% faster, COCO128 ~3.6% slower), not a universal speedup. Reducing `workers` and/or image/batch size is a current memory-pressure workaround. The remaining issue is bounded resource use under the reported configuration, not simply that a persistent worker is alive.

---

### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train, Val

### General Summary

During detection training, validation uses an `InfiniteDataLoader`. Its worker processes and prefetched batches remain alive after a validation pass completes.

Detection validation also requests twice the configured worker count. With large images, this can retain tens of gigabytes of host memory, saturate the CPU with memory reclaim/compaction, and leave the GPU idle between epochs. There is no exception or traceback. The failure mode is excessive resource retention.

### Actual Behavior

For the following training configuration:

* workers=15
* batch=16
* imgsz=1920
* cache=False
* 20 host CPUs
* NVIDIA GB10 with unified memory
    I observed:
* 15 training pt_data_worker processes
* 20 validation workers, because validation requests workers * 2, capped by the host CPU count
* Approximately 95–100 GB total memory usage
* 100% CPU utilization after validation completed
* kcompactd0 consuming a full CPU core
* GPU mostly idle during the resulting memory pressure
* Validation worker processes still alive after validation exited

Memory Breakdown

The training parent process had approximately:

* 38.6 GiB PSS
* Approximately 36.8 GiB shared-memory PSS
* Approximately 29.8 GiB NVIDIA allocation
    Representative training workers consumed approximately 1.7 GiB PSS each.
    This occurred with cache=False, so the memory usage was not caused by dataset caching.

### Validation Prexetch Memory

At a validation batch size of 32, a raw uint8 image batch at 1920×1920 is approximately 337.5 MiB per batch.
With 20 workers and a prefetch factor of 4, validation may queue as many as 20 workers × 4 batches/worker = 80 batches.

This corresponds to approximately 80 × 337.5 MiB ≈ 27 GiB of raw image data alone, before accounting for multiprocessing, shared-memory, tensor, and Python overhead.
The combination of persistent validation workers and aggressive prefetching therefore creates substantial retained memory pressure even after the validation pass has completed.

### Expected Behavior

* Training may continue using InfiniteDataLoader and persistent workers.
* Validation should use a normal finite `torch.utils.data.DataLoader`.
* Validation workers and prefetch queues should terminate after validation.
* workers=N should not unexpectedly create 2N validation workers.
* Validation should use a conservative prefetch factor to avoid large queued batches when training with high-resolution images.

### Environment

 ```text
  Ultralytics 8.4.120
  Python 3.12.13
  PyTorch 2.13.0+cu130
  Torchvision 0.28.0
  CUDA 13.0

  OS          Linux 6.17.0-1029-nvidia, aarch64
  CPU count   20
  RAM         121.7 GB
  GPU         NVIDIA GB10, 124610 MiB
  GPU count   1

  numpy                 1.26.4
  opencv-python         4.11.0.86
  pillow                12.3.0
  psutil                7.2.2
  polars                1.43.2
  ultralytics-thop      2.1.6
  ultralytics-platform  0.1.5

### Minimal Reproducible Example

  ```python
  from ultralytics.data.build import build_dataloader

  loader = build_dataloader(
      range(64),
      batch=8,
      workers=2,
      shuffle=False,
      device="cpu",
  )

  workers = list(loader.iterator._workers)

  try:
      print("loader type:", type(loader).__name__)
      list(loader)

      alive = [worker.is_alive() for worker in workers]
      print("workers alive after full iteration:", alive)
      print("worker PIDs:", [worker.pid for worker in workers])

      assert not any(alive), "Workers remain alive after the finite pass"
  finally:
      loader.close()
```

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @danielmorozoff, thank you for the detailed report. The finite validation pass retaining worker processes and prefetched batches—along with validation requesting twice the configured worker count—is a concrete resource-management issue, and the reproduction plus memory measurements are very helpful.

Your proposed direction of using a finite validation `DataLoader`, ensuring workers terminate after validation, and avoiding unexpectedly doubled worker counts is well scoped. If you submit a PR, please include a regression test that verifies validation workers are no longer alive after a finite pass and that the configured worker count is respected.

This is an automated response, and an Ultralytics engineer will review the report and assist soon.

**Comment by Y-T-G** [comment]
Thanks for reporting and opening a PR.

The worker count is intended behavior and so is `prefetch_factor`. So those changes will be rejected.

The cleanup is the actual bug here.

**Comment by danielmorozoff** [comment]
I understand that worker reuse and the current defaults are intentional. The concern is that there is currently no supported mechanism to bound validation prefetching or disable validation worker persistence for memory-constrained workloads.

In my example, at imgsz=1920, memory remains at approximately 95–100 GB after validation, even with a single YOLO26n model. With 20 validation workers and prefetch_factor=4, the raw prefetch queue alone has an estimated upper bound of approximately 27 GiB, before multiprocessing, shared-memory, tensor, and Python overhead.

This triggers memory compaction, saturates the CPU, and leaves the GPU idle. At that point, worker persistence reduces throughput rather than improving it. Keeping the existing defaults is reasonable, but users need a supported way to either bound validation prefetching or release validation workers between passes.

PyTorch supports the setting, but Ultralytics’ wrapper does not expose it. 

#### See below for top profiles (during validation) and notice how the GPU just bursts and the CPUs are fully saturated even at 75% workers setting in the yaml file :

<img width="963" height="331" alt="Image" src="https://github.com/user-attachments/assets/f2969588-6455-4224-9889-7c5e2117a066" />


<img width="433" height="102" alt="Image" src="https://github.com/user-attachments/assets/930cf5c2-7ff1-403a-b434-97d8883a7e22" />


**The CPU is using 1 core for compaction (top process) + then the validation workers don't get cleaned up and memory persists (this happens in the training step for validation workers and then for training workers during validation)**

#### Here is the same model during the training (non validation phase) once it stabilizes -- see below:

<img width="970" height="317" alt="Image" src="https://github.com/user-attachments/assets/cd516a73-c083-49e8-a7b1-a81989dee01f" />


<img width="418" height="119" alt="Image" src="https://github.com/user-attachments/assets/4d7ef4e5-6d15-401e-bc4c-54c26587338d" />


Setting | Training | Validation
-- | -- | --
Batch size | 16 | 32
Workers | 15 | 20 (workers * 2, CPU-capped)
Prefetch per worker | 4 batches | 4 batches
Queue upper bound | 60 batches | 80 batches
Approx. raw-image queue | 9.9 GiB | 26.4 GiB

### Total utilization during validation:
This is a unified memory system so the model is 36.8GiB and the operations overhead are the workers / compaction processes

Component | PSS
-- | --
Parent training process | 36.80 GiB
15 training workers | 19.51 GiB
20 validation workers | 13.87 GiB
Total training job | 70.18 GiB


**Note that there is an intermediate phase that follows transition from validation back to training::**
This continued  about 100 steps into training...

When validation returned, its InfiniteDataLoader did not cancel work already assigned to the 20 validation workers. Because the sampler repeats indefinitely, up to roughly 80 validation batches for the next pass remained outstanding.

Those workers continued in the background while training ran for another 98 iterations:

  1. Validation returned.
  2. Training resumed.
  3. **Validation workers kept decoding and assembling outstanding batches <-- (useless work)**
  4. Training and validation workers competed for memory.
  5. The kernel continued compaction.
  6. Once the validation result queue filled, its workers blocked and CPU pressure collapsed.

<img width="973" height="333" alt="Image" src="https://github.com/user-attachments/assets/71b0aae5-544b-4ee1-88bd-20d70f62a01c" />


Also here is system background performance without the Ultralytics training loop running:

<img width="964" height="328" alt="Image" src="https://github.com/user-attachments/assets/3716fde4-45de-4bec-a341-980c5019bf68" />

<img width="430" height="182" alt="Image" src="https://github.com/user-attachments/assets/f4a2998c-a0b8-4d59-a2d8-8707dbe8a9bc" />

**Comment by danielmorozoff** [comment]
### Proposed fix,  profile during transition #25828 

This is the transition from training to validation... (Note this is another model)

<img width="977" height="324" alt="Image" src="https://github.com/user-attachments/assets/318c2294-00e1-4717-9eb4-40276e6c6d99" />


<img width="426" height="113" alt="Image" src="https://github.com/user-attachments/assets/00b7fefa-fc4f-4ed8-beee-eba8d9ba2267" />


and then back to training:

<img width="973" height="316" alt="Image" src="https://github.com/user-attachments/assets/771ce51c-d182-46b7-b550-ad96f4861bd4" />


<img width="435" height="114" alt="Image" src="https://github.com/user-attachments/assets/d7fcded7-e406-43c4-a66c-64762a96ddec" />

**Comment by Y-T-G** [comment]
> The concern is that there is currently no supported mechanism to bound validation prefetching or disable validation worker persistence for memory-constrained workloads.

We unfortunately can't update default behavior of the library because of a single user's constraints. Ultralytics is used by hundreds of thousands of users. We also can't provide controls for every behavior in Ultralytics since that would bloat the library. Ultralytics is open source, and you can modify the behavior if needed. You could probably even modify it through trainer inheritance without modifying source code.

> This triggers memory compaction, saturates the CPU, and leaves the GPU idle. At that point, worker persistence reduces throughput rather than improving it. Keeping the existing defaults is reasonable, but users need a supported way to either bound validation prefetching or release validation workers between passes.

Do you have any comparison of training time before and after your fix? I don't think this behavior is much different from `persistent_workers=True`. Ultralytics implements its own version of it which is faster. Releasing and recreating workers introduce their own overhead. So the fix simply trades one overhead for another.

> Validation workers kept decoding and assembling outstanding batches <-- (useless work)

It's not really useless work. Because the prefetched batches are what would be used as soon as next validation starts. Whilst with the fix, it would need to first recreate those workers, then prefetch the first few batches before returning something

**Comment by Y-T-G** [comment]
If you're interested, you could contribute a new feature that dynamically scales worker count to fit device memory.

That would be better for every user as opposed constraining the default behavior to lowest common denominator.

**Comment by danielmorozoff** [comment]
This will be on my back burner if I get to pick my head up from the day job.

**Comment by cainiao33** [comment]
Happy to pick this up — I ran measurements first so the design discussion starts from real numbers.

**Measured worker footprint** (yolo26n, COCO128, Ryzen 7 5700X / RTX 4060 Ti 16 GB, PSS via `smaps_rollup` — same accounting as the tables above; train and val loaders coexist since `test_loader` is built upfront, and val's batch×2 / prefetch-2 gives each val worker the same queue bytes as a train worker):

| workers×batch×imgsz | workers alive | total worker PSS | per-worker PSS | epoch time |
|---|---|---|---|---|
| 8×8×640 | 16 | 2.3 GiB | ~146 MiB | 34 s* |
| 8×16×320 | 12 | 1.7 GiB | ~143 MiB | 25 s* |
| 8×16×640 | 12 | 2.5 GiB | ~205 MiB | 28 s* |
| 8×16×1280 | 12 | 4.7 GiB | ~393 MiB | 87 s* |
| 8×32×640 | 6 | 2.0 GiB | ~332 MiB | 41 s* |
| 2/4/8/16×16×640 | 6/8/12/12 | 1.8–2.6 GiB | — | 13.9/12.5/12.4/12.5 s |

\* single-epoch runs include startup; the last row is 3-epoch averages.

**Model**: per-worker PSS ≈ `124 MiB + 0.93 × (prefetch_factor × batch × imgsz² × 3 B)` — fits all five batch/imgsz configs within ±8%, and sanity-checks against this thread's report (15 train workers @ batch 20 / imgsz 1920 → predicts ~14 GiB vs 19.5 GiB measured, same order; larger images/labels there). RSS sums are 4–9× higher than PSS due to fork-shared pages, so PSS is the metric to budget on.

**Throughput knee**: 2→4 workers gives ~10% faster epochs; 8→16 gives nothing. Memory-bounding the worker count costs almost no throughput.

**Proposal**:

- UX: `workers=-1` opts in, mirroring `batch=-1` → AutoBatch. Default stays 8, no change for existing configs.
- Resolution in `build_dataloader()` (already receives `device` and `batch`, computes `prefetch_factor`, applies the existing `min(cpu // nd, workers, batches)` caps): `workers_auto = clamp(available_RAM × safety / (C + k × pf × batch × imgsz² × 3B), 1, cpu // nd, batches)`
- Workers budget **host** RAM via `psutil.virtual_memory().available` (psutil is already a dependency); on unified-memory systems that pool *is* the device memory, so one reading covers both — measured worker VRAM contribution is flat regardless of worker count, the queues never live in VRAM.

Three questions before I start:

1. Opt-in `workers=-1` first, with auto-as-default as a possible later step once proven?
2. Host-RAM budgeting as above, or do you want the accelerator `mem_get_info()` reading even on discrete GPUs?
3. Log the resolved count AutoBatch-style (`AutoWorkers: N workers`)?

Happy to rebase on #26014 — the count-resolution change composes with its loader-lifetime work.

**Comment by pderrenger** [comment]
Please start with opt-in `workers=-1`, keep the existing default unchanged, budget against available host RAM rather than accelerator memory, and log the resolved count as `AutoWorkers: N workers`. Keep this as a separate minimal PR from #26014, avoid changes to `default.yaml`, and include focused tests and docs for clamping, DDP, tiny datasets, and low-memory fallback.

**Comment by cainiao33** [comment]
Implemented along these lines — opened #26090. The prototype measurements confirmed all three choices above. One addition that came out of benchmarking: the memory budget is applied as a ceiling that only **lowers** the default count, never raises it — a pure budget-based resolution tops out at the cpu cap (e.g. 16 workers), which ran 15–29% slower per epoch than the default at imgsz 640/1280, since beyond the throughput knee extra workers are pure process overhead. With the ceiling-at-default design memory-rich hosts keep exactly the current behavior and constrained hosts get a safe value.

Thanks for the review guidance: no `default.yaml` changes, focused tests for clamping / DDP resolution / tiny datasets / low-memory floor, and kept fully independent of #26014.

**Comment by pderrenger** [comment]
Thanks for opening #26090. The ceiling-at-default design is aligned with the intended opt-in behavior and avoids increasing process overhead on memory-rich hosts; we’ll continue review on the PR, including the focused tests, docs, and benchmark evidence.
